### PR TITLE
Use bootstrap4

### PR DESCRIPTION
**Pull request by OGAWAHirofumi**
This migrate bootstrap3 to bootstrap4. With this, IMO, becomes better views (but im not a web developer).

however, testing and checking UI is hard (no auto test, and people has own favorite), so this PR would be needed carefully review more or less.

### CODE REVIEW

**Comment by ThomasWaldmann** [comment]
Can you do a **minimal** change please to upgrade to bs4?

**Comment by OGAWAHirofumi** [comment]
> Can you do a **minimal** change please to upgrade to bs4?

minimal means, doesn't care bad view by convert? e.g. looks like some css class (e.g. page-header) was removed from bootstrap4.

**Comment by ThomasWaldmann** [comment]
Well, I see some changes that do not look minimal, like:

Size changes  `sm -> md` (small to medium)?

Colours are being added here and there, why?

Some stuff looks like new features, like:
- Use shorter name for "Create List Item"
- Use full width for navbar like bootstrap4 examples

In a pull request that upgrades bootstrap to a new version, there should be only changes that make the UI work with bootstrap 4.

Regression fixes related to that upgrade are also ok (so it does not look obviously bad / wrong **if** it looked ok with bs3).

But other than that, there should be no other changes **within this PR**.

**Improvements on looks and usability are welcome in general, but should be separate.**

**Comment by OGAWAHirofumi** [comment]
> Well, I see some changes that do not look minimal, like:
> 
> Size changes `sm -> md` (small to medium)?

col-<size> were changed in bootstrap4. like xs => default, sm => md.

> Colours are being added here and there, why?

<component>-default was removed.

> Some stuff looks like new features, like:
> 
>     * Use shorter name for "Create List Item"
> 
>     * Use full width for navbar like bootstrap4 examples
> 
> 
> In a pull request that upgrades bootstrap to a new version, there should be only changes that make the UI work with bootstrap 4.
> 
> Regression fixes related to that upgrade are also ok (so it does not look obviously bad / wrong **if** it looked ok with bs3).
> 
> But other than that, there should be no other changes **within this PR**.
> 
> **Improvements on looks and usability are welcome in general, but should be separate.**

ah, it was why separated patches (can revert easily). but you wanted separated **PR** somewhat. ok, will separate.

**Comment by OGAWAHirofumi** [comment]
Added screen shot script to see UI changes.

[screenshots-before.tar.gz](https://github.com/bepasty/bepasty-server/files/5500373/screenshots-before.tar.gz)
[screenshots-after.tar.gz](https://github.com/bepasty/bepasty-server/files/5500375/screenshots-after.tar.gz)

screenshots-before == bootstrap3
screenshots-after == bootstrap4

those should make easier to review.

with result of screenshots, adjusted button wrap in display view, and tweaked QR view to be more bootstrap3

File: src/bepasty/tests/screenshots.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+        self.wait_present("//input[@value='Logout']")
+
+        # NOTE: upload screen, 1 - close hamburger, 2 - open hamburger
+        self.top_screen_shots("03-upload")
+
+        try:
+            self.browser.find_element_by_xpath("//input[@value='Logout']")
+        except NoSuchElementException:
+            raise ValueError("Can't login!!! Create a user 'foo' with the permissions"
+                             "'read' and 'create' in your PERMISSIONS in the config")
```
Comment (ThomasWaldmann) [inline src/bepasty/tests/screenshots.py:114]: not "user", it is a "secret".

So it becomes:

Can't login! Please edit your config, go to PERMISSIONS setting and add a new secret 'screenshots' with 'read' and 'create' permissions.

**Comment by ThomasWaldmann** [comment]
Looks good! One minor issue to fix and it's ready to merge!

**Comment by ThomasWaldmann** [comment]
Guess I need to update the bootstrap 4 xstatic package ...

**Comment by OGAWAHirofumi** [comment]
changed

- tweaked screenshots.py error message to say for 'foo' and all permissions
- use xstatic-bootstrap>=4.0.0.0,<5.0.0.0
- screenshots.tar.gz with bootstrap 4.5.1 (Create List Item button wrap now, this will fix separatedly)

[screenshots.tar.gz](https://github.com/bepasty/bepasty-server/files/5505870/screenshots.tar.gz)

**Comment by OGAWAHirofumi** [comment]
tested with latest xstatic stuff

only noticed change looks like `<pre>` color in pygment (background color for line number)

[screenshots.tar.gz](https://github.com/bepasty/bepasty-server/files/5505930/screenshots.tar.gz)

**Comment by OGAWAHirofumi** [comment]
tested with latest xstatic stuff. only noticed change looks like `<pre>` color in pygment (background color for line number)

[screenshots.tar.gz](https://github.com/bepasty/bepasty-server/files/5505948/screenshots.tar.gz)

changed

- improved screenshots.py to take screen shot for bootbox

previous

- tweaked screenshots.py error message to say for 'foo' and all permissions
- use xstatic-bootstrap>=4.0.0.0,<5.0.0.0
- screenshots.tar.gz with bootstrap 4.5.1 (Create List Item button wrap now, this will fix separatedly)

**Comment by OGAWAHirofumi** [comment]
tested with latest xstatic stuff. only noticed change looks like `<pre>` color in pygment (background color for line number)

[screenshots.tar.gz](https://github.com/bepasty/bepasty-server/files/5506071/screenshots.tar.gz)

changed

- tweaked screen height for screenshots.py
- improved screenshots.py to take screen shot for bootbox
- added line highlight for display view screen shot

previous

- tweaked screenshots.py error message to say for 'foo' and all permissions
- use xstatic-bootstrap>=4.0.0.0,<5.0.0.0

File: setup.py [hunk 2]
```diff
@@ -43,7 +43,7 @@
         'xstatic',
         'XStatic-asciinema-player',
         'xstatic-bootbox',
-        'xstatic-bootstrap<4.0.0.0',
+        'xstatic-bootstrap>=4.0.0.0,<5.0.0',
```
Comment (ThomasWaldmann) [inline setup.py:46]: please use 4 digits for both version numbers

File: src/bepasty/tests/screenshots.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+        self.wait_present("//input[@value='Logout']")
+
+        # NOTE: upload screen, 1 - close hamburger, 2 - open hamburger
+        self.top_screen_shots("03-upload")
+
+        try:
+            self.browser.find_element_by_xpath("//input[@value='Logout']")
+        except NoSuchElementException:
+            raise ValueError("Can't login! Please edit your config, go to PERMISSIONS setting "
+                             "and add a new secret 'foo' with all permissions.")
```
Comment (ThomasWaldmann) [inline src/bepasty/tests/screenshots.py:114]: can you use a better secret than 'foo' here? maybe 'for-screenshots'?

File: setup.py (hunk 2 above)
Comment (OGAWAHirofumi) [reply setup.py:46]: ouch, will fix

File: src/bepasty/tests/screenshots.py (hunk 3 above)
Comment (OGAWAHirofumi) [reply src/bepasty/tests/screenshots.py:114]: 'foo' is default entry in config.py. and it is what test_website.py uses too.
changing from 'foo', just will make inconvenient to test.

**Comment by OGAWAHirofumi** [comment]
tested with latest xstatic stuff. only noticed change looks like `<pre>` color in pygment (background color for line number)

[screenshots.tar.gz](https://github.com/bepasty/bepasty-server/files/5506071/screenshots.tar.gz)

changed

- use 5.0.0.0

previous

- tweaked screen height for screenshots.py
- improved screenshots.py to take screen shot for bootbox
- added line highlight for display view screen shot
- tweaked screenshots.py error message to say for 'foo' and all permissions
- use xstatic-bootstrap>=4.0.0.0,<5.0.0.0
- screenshots.tar.gz with bootstrap 4.5.1 (Create List Item button wrap now, this will fix separatedly)

File: src/bepasty/tests/screenshots.py (hunk 3 above)
Comment (ThomasWaldmann) [reply src/bepasty/tests/screenshots.py:114]: oh, ok. so keep it for now.

**Comment by ThomasWaldmann** [comment]
OK, looks like this is finished! Will merge...

**Pull request merged**